### PR TITLE
Resolves #279: Add instructions to the circuit breaker guide

### LINKED ISSUE #279: Add instructions to the Circuit Breaker section of the Dev Guides
Resolved by PR #290

### PR DESCRIPTION
**Pull request by inoxx03**
Instructions cleaned up edited with the review suggestions factored in.
@Ladicek and @rhoads-zach Mind taking a look?

### CODE REVIEW

File: docs/topics/circuit-breaker-mission-intro.adoc [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
-With the Circuit Breaker pattern, a service client invokes a remote service at regular time intervals. If the call to the remote service fails repeatedly, and the number of consecutive failures exceeds a specified limit, the circuit breaker makes the service unavailable for a specified time period, thus preventing any remote calls from other services from reaching the failed service. once the timeout period expires, the circuit breaker allows a limited number of test calls to pass through to the remote service. If these test calls result in failure, the timeout period restarts and the service remains unavailable. If the calls succeed, the Circuit Breaker makes the service available to remote calls by other services.
+In an architecture where multiple services depend on each other for functionality, a failure in one service, can rapidly propagate to its dependent services, causing the entire architecture to collapse. Implementing a Circuit Breaker pattern helps prevent this.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-intro.adoc]: I guess "a failure in one service, can rapidly propagate" contains an extra comma.

**Comment by Ladicek** [comment]
The general text about circuit breakers LGTM.

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+. Use `curl` to execute a `GET` request against the `{app-name}` service. You can also use a browser to do this.
++
+[source,bash,options="nowrap",subs="attributes"]
+----
+$ curl http://{app-name}-{project-name}.1ab5.starter-us-east-1.openshiftapps.com/api/greeting
+{"content":"Hello, World!"}
+----
+
+. Invoke the /api/killme endpoint and verify the availability of the /api/greeting endpoint shortly after that.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: There's no `/api/killme` endpoint. It's `/api/state`, the correct HTTP verb is `PUT` and it expects a JSON request body, as described in the mission definition.

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
++
+Invoking the /api/killme endpoint simulates an internal service failure and triggers the OpenShift self-healing capabilities. When invoking /api/greeting after simulating the failure, the service should return a HTTP status 503.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: It _doesn't_ trigger OpenShift self-healing capabilities. It seems this is copied from the Health Check docs, but that's obviously something completely different :-)

**Comment by Ladicek** [comment]
Please ignore the 2 comments I added (and then deleted), I was looking at the individual commit instead of the final state.

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
++
+[source,bash,options="nowrap",subs="attributes"]
+----
+$ curl http://{app-name}-greeting-{project-name}.1ab5.starter-us-east-1.openshiftapps.com/api/greeting
+{"content":"Hello, World!"}
+----
++
+// Add note about the Toggle button not working
+// no scaler implemented error if CLI used to scale down pod
+. To simulate the failure of the `{app-name}-name` service, use the `Toggle` button in the web interface. Alternatively, you can scale the pod running the `{app-name}-name` service down to 0 replicas using the web interface, or by executing the following command.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: Or you can do something like `curl -X PUT -H "Content-Type: application/json" -d '{"state": "fail"}' http://.../api/state`.

To recover, just change it to `{"state": "ok"}`.

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
++
+. The indicator of the circuit breaker state in the web interface changes from `CLOSED` to `OPEN` and the Circuit breaker should issue the fallback response when you invoke the `api/greeting` endpoint.
++
+[source,bash,option="nowrap",subs="attributes+"]
+----
+$ curl http://{app-name}-greeting-{project-name}.1ab5.starter-us-east-1.openshiftapps.com/api/greeting
+{"content":"Hello, Fallback!"}
+----
++
+. Scale the pod running the `{app-name}-name` service back up to 1 or wait for it to heal itself and invoke the `api/greeting` endpoint again. If the service becomes available, you should receive the `Hello World` greeting response.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: It will never heal itself, you have to do that manually in this mission.

File: docs/topics/circuit-breaker-mission-deploy-booster.adoc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+. Create a new project in your {OpenShiftOnline}.
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+$ oc new-project {project-name}
+----
+
+. Navigate to the root directory of your booster.
+
+. Use maven to start the deployment to your {OpenShiftOnline}.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-deploy-booster.adoc]: maven -> Maven

**Comment by inoxx03** [comment]
@Ladicek Thanks for the extensive feedback:) I've made some changes based on your comments and also added some detail to the endpoint description. Please take a look whenever you have time.

**Comment by Ladicek** [comment]
Hi @inoxx03, I see some factual issues in this PR. I felt it would be too hard for me to describe them in comments, so I instead modified the Spring Boot part myself -- here's the commit: https://github.com/Ladicek/appdev-documentation/commits/cb-fixes

In addition to fixing the facts, I also modified the structure a little -- I moved the description of the `name` service _before_ the description of the `greeting` service, because the latter refers to the former. Feel free to improve the wording or do any other changes you deem necessary. The Vert.x and Swarm parts are basically the same as the Spring Boot one, so maybe the text could be shared among all the runtimes.

**Comment by rhoads-zach** [comment]
thanks @inoxx03 merging.

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+* the `/api/state` endpoint, which controls the behavior of the `/api/name` endpoint and determines whether the service works correctly or demonstrates failure.
+
+`{app-name}-greeting`::
+Exposes the following endpoints:
+
+* the `/api/greeting` endpoint that you can call to get a personalized greeting response.
++
+Calling the `/api/greeting` endpoint also calls the `/api/name` endpoint of the `{app-name}-name` service.
+This call happens behind the scenes and is protected by the Circuit Breaker.
```
Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: I agree my original wording wasn't terribly awesome here. What I wanted to convey by the "behind the scenes" words is: as part of processing of the `/api/greeting` endpoint, the `{app-name}-greeting` service calls the `/api/name` endpoint of the `{app-name}-name` service. This call is guarded by the circuit breaker.

To me, the sentence "This call happens behind the scenes and is protected by the Circuit Breaker" places an equal emphasis on "happens behind the scenes" and "is protected by the circuit breaker", which isn't really appropriate. The really important thing here is that the call is guarded by the circuit breaker. The fact that `{app-name}-greeting` calls into `{app-name}-name` is just a description of how the services interact.

IDK, can we improve the wording somehow? Maybe I'm just nitpicking here :-)

**Comment by rhoads-zach** [comment]
merged: https://github.com/openshiftio/appdev-documentation/commit/313b765597e327d462b1d3ca0de41635aa338b5f

File: docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc (hunk 7 above)
Comment (rhoads-zach) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: @Ladicek sorry, just saw this and already merged. @inoxx03 feel free to file a follow up PR and i'll merge it.

Comment (Ladicek) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: No problem at all! I just remembered to check this when I saw your "merging" comment, so thanks! :-)

Comment (inoxx03) [inline docs/topics/circuit-breaker-mission-basic-interaction-spring-boot-tomcat.adoc]: @Ladicek I think you make a good point in your comment. Let me rephrase that and file a followup PR:)

**Comment by inoxx03** [comment]
@rhoads-zach Thanks for merging:)
